Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,265 @@
+
+    public BocListValidationFailureRepository ()
+    {
+      _listValidationFailures = new List<HandleableFailure>();
+      _rowValidationFailures = new Dictionary<IBusinessObject, RowFailureContainer>();
+    }
+
+    public void AddListValidationFailures (IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      _listValidationFailures.AddRange(validationFailures.Select(e => new HandleableFailure(e)));
+    }
+
+    public void AddRowValidationFailures (IBusinessObject rowObject, IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(rowObject), rowObject);
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      var failureContainer = GetOrCreateRowFailureContainer(rowObject);
+      failureContainer.AddRowFailures(validationFailures);
+    }
+
+    public void AddCellValidationFailures (IBusinessObject rowObject, IBocColumnDefinitionWithValidationSupport columnDefinition, IEnumerable<BusinessObjectValidationFailure> validationFailures)
+    {
+      ArgumentUtility.CheckNotNull(nameof(rowObject), rowObject);
+      ArgumentUtility.CheckNotNull(nameof(columnDefinition), columnDefinition);
+      ArgumentUtility.CheckNotNull(nameof(validationFailures), validationFailures);
+
+      var failureContainer = GetOrCreateRowFailureContainer(rowObject);
+      failureContainer.AddCellFailures(columnDefinition, validationFailures);
+    }
+
+
+    public IReadOnlyCollection<BusinessObjectValidationFailure> GetValidationFailuresForBocList (bool markAsHandled)
+    {
+      return GetUnhandledValidationFailuresForBocList(markAsHandled);
+    }
+
+    public IReadOnlyCollection<BusinessObjectValidationFailure> GetValidationFailuresForBocListAndContainingDataRowsAndDataCells (bool markAsHandled)
+    {
+      var failures = GetUnhandledValidationFailuresForBocList(markAsHandled);
+
+      foreach (var failureContainer in _rowValidationFailures.Values)
+      {
+        var rowAndCellFailures = failureContainer.GetValidationFailuresForDataRowAndDataCells().Where(f => !f.Handled);
+        foreach (var handleableFailure in rowAndCellFailures)
+        {
+          failures.Add(handleableFailure.Failure);
+          if (markAsHandled)
+            handleableFailure.MarkAsHandled();
+        }
+      }
+
+      return failures;
+    }
+
+    public IReadOnlyCollection<BusinessObjectValidationFailure> GetValidationFailuresForDataRow (IBusinessObject rowObject, bool markAsHandled)
+    {
+      ArgumentUtility.CheckNotNull(nameof(rowObject), rowObject);
+
+      var failureContainer = GetRowFailureContainer(rowObject);
+      if (failureContainer == null)
+        return Array.Empty<BusinessObjectValidationFailure>();
+
+      var rowFailures = failureContainer.GetValidationFailuresForDataRow().Where(f => !f.Handled).ToList();
```

Comment:
This would create multiple enumeration, is that preferred here?

Reply:
I see, the next comment fixes that.

Reply:
Correct and true. We never want multiple enums, sometimes it's a detail that gets missed in review suggestions :) Anyway, since the updated design solves this anyway, the point is moot :)